pulp_docker/app/viewsets.py

```diff
@@ -280,3 +285,46 @@ def create(self, request):
             }
         )
         return OperationPostponedResponse(result, request)
+
+
+class TagCopyViewSet(drf_viewsets.ViewSet):
+    """
+    ViewSet for copying tags recursively.
+    """
+
+    serializer_class = serializers.TagCopySerializer
+
+    @swagger_auto_schema(
+        operation_description="Trigger an asynchronous task to copy tags",
+        responses={202: AsyncOperationResponseSerializer},
+        request_body=serializers.TagCopySerializer,
+    )
+    def create(self, request):
+        """
+        Queues a task that creates a new RepositoryVersion by adding content units.
+        """
+        names = request.data.get("names")
+        serializer = serializers.TagCopySerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        source = serializer.validated_data['source_repository']
+        destination = serializer.validated_data['destination_repository']
+        source_latest = RepositoryVersion.latest(source)
+        content_tags_in_repo = source_latest.content.filter(
+            _type="docker.tag"
+        )
+        tags_in_repo = models.Tag.objects.filter(
+            pk__in=content_tags_in_repo,
+        )
+        if names is None:
+            tags_to_add = tags_in_repo
+        else:
+            tags_to_add = tags_in_repo.filter(name__in=names)
+
+        result = enqueue_with_reservation(
+            tasks.recursive_add_content, [destination],
```

Comment:
i think we should also lock source_repo, especially if it looks at latest version, which can change

Reply:
On second thought, there is no need to reserve on the source repo. The viewset filters the content of the source repository synchronously, and the recursive_add task is given a list of content hrefs, which are independent of the source repo.